Feat: Backend for Questionnaires

haanh764 opened the pull request:
Hello, this PR covers the requirements for questionnaires in backend side XD
Notes: I have changed the old module `question` (I created before that intended to use for handling the questions in affective annotation project but then unused) to `questionnaires`

What's new:
- `backend/config/settings/base.py`: added new application in config `questionnaires`
- `backend/config/urls.py`: added new app's urls to config 
- `backend/questionnaires/apps.py`: added new `questionnaires` app
- `backend/questionnaires/models.py`: added new models for questionnaires, including `questionnaire_type`, `questionnaires`, `questions`, `answers`
- `backend/questionnaires/migrations/0002_initial.py`, `backend/questionnaires/migrations/0003_initial.py`: added migration due to changes in models, populate data for 3 tables `questionnaire_type`, `questionnaires`, `questions`
- `backend/questionnaires/serializers.py`: added serializers for these new models
- `backend/questionnaires/views.py`: added views for these new models
- `backend/questionnaires/urls.py`: added endpoints for these new models
- `backend/questionnaires/test/test_views.py`: added test cases for these new views

What's changed:
- `backend/labels/migrations/0015_textlabel_question.py` - `backend/labels/migrations/0022_remove_scale_question.py`: adjusted the migration files due to the changes when I removed the unused module `questions` (since some of them have dependencies, for example when I added or removed the question references before)
- `backend/questions/_`: removed unused module `questions`
- `backend/pyproject.toml`: added new app

The endpoints and format of responses are as below:
- List all types of questionnaires: `/v1/questionnaire_types`
```
{
    "count": 6,
    "next": "http://localhost:8000/v1/questionnaire_types?limit=5&offset=5",
    "previous": null,
    "results": [
        {
            "id": 1,
            "name": "Przed badaniem",
            "description": ""
        },
        {
            "id": 2,
            "name": "Przed i po badaniu",
            "description": ""
        },
        {
            "id": 3,
            "name": "Po pierwszym tygodniu i na końcu badań",
            "description": ""
        },
        {
            "id": 4,
            "name": "Codziennie w trakcie badania",
            "description": ""
        },
        {
            "id": 5,
            "name": "Ankieta na koniec badania",
            "description": ""
        }
    ]
}
```
- List all questionnaires of given type: `/v1/questionnaires/<int> questionnaire_tye_id (for eg type 1)`
```
{
    "count": 3,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": 1,
            "name": "IPIP-BFM-20 (Osobowość)",
            "description": "",
            "questionnaire_type": 1,
            "language": "pl",
            "time_of_day": ""
        },
        {
            "id": 2,
            "name": "Demografia",
            "description": "",
            "questionnaire_type": 1,
            "language": "pl",
            "time_of_day": ""
        },
        {
            "id": 3,
            "name": "Humor Styles Questionnaire",
            "description": "",
            "questionnaire_type": 1,
            "language": "pl",
            "time_of_day": ""
        }
    ]
}
```
- List all questions of given questionnaire: `v1/questionnaires/<int>questionnaire_id (for eg 1)/questions`
```
{
    "count": 20,
    "next": "http://localhost:8000/v1/questionnaires/1/questions?limit=5&offset=5",
    "previous": null,
    "results": [
        {
            "id": 1,
            "question_text": "Jestem duszą towarzystwa.",
            "questionnaire": 1
        },
        {
            "id": 2,
            "question_text": "Niezbyt obchodzą mnie inni ludzie.",
            "questionnaire": 1
        },
        {
            "id": 3,
            "question_text": "Zostawiam moje rzeczy gdzie popadnie.",
            "questionnaire": 1
        },
        {
            "id": 4,
            "question_text": "Zwykle jestem zrelaksowany/a.",
            "questionnaire": 1
        }
.....
}
```
- Create answer: `answers/<int:question_id>`, format to send request for eg:
```
{   
    "answer_text": "finally it's friday!!!!",
    "question": 15
}
```

Edit: I have changed the endpoint for create answer. The user flow to answer/change their answer is as below:
- Query the user's answer by `question_id` and `created_date` (check if that answer is created today). 
- If for a question, a user have answer it today (there's existed answer in dtb), then he/she can update the answer, otherwise a new answer will be added.

teddy-f-47 reviewed (commented):
Hello, thank you for the nice work! Overall I think the code looks fine. I just have a few questions:
1. With the `AnswerListAPI`, each answer to a question must be submitted individually (one POST for each answer in the questionnaire), right? I think this is the expected behavior, just asking to make sure XD
2. With the `AnswerListAPI`, is it possible for a user to submit an answer, delete it, and submit a new one? If not, how does a user edit an existing answer?
3. When I try to run the tests with `poetry run task test`, poetry seems to be unable to find the new test files from `questionnaires` module. In the original affective annotation branch it found 483 tests, and in this branch it also found 483 tests. I also tried print() stuff in the new test file, but the prints are not displayed in the cmd. Would you please check it? XD

teddy-f-47 commented:
3. I think I have found the solution: (1) we need to add an empty `__init__.py` file in questionnaires/test/ folder, and (2) we need to import `from api.tests.utils import CRUDMixin` inside `questionnaires\test\test_views.py` and make all the classes inherit from CRUDMixin (e.g. `class TestQuestionnaireTypeList(CRUDMixin)`) 😅 Now poetry can find the new tests in the new test file, but some tests fail😓

haanh764 commented:
> Hello, thank you for the nice work! Overall I think the code looks fine. I just have a few questions:
> 
> 1. With the `AnswerListAPI`, each answer to a question must be submitted individually (one POST for each answer in the questionnaire), right? I think this is the expected behavior, just asking to make sure XD
> 2. With the `AnswerListAPI`, is it possible for a user to submit an answer, delete it, and submit a new one? If not, how does a user edit an existing answer?
> 3. When I try to run the tests with `poetry run task test`, poetry seems to be unable to find the new test files from `questionnaires` module. In the original affective annotation branch it found 483 tests, and in this branch it also found 483 tests. I also tried print() stuff in the new test file, but the prints are not displayed in the cmd. Would you please check it? XD

1. Yes, it is submitted individually XD
2. Yeah, I forgot that maybe users will need to change their answer. I will update it XD
3. Okie I will check it also, probably I forgot to run the test to check it again XD

artikandri commented:
Hello, I have tried implementing some of the apis and while I can't confirm that it works well in the whole cycle since I havent finished everything yet, I can confirm that it works XD  but I got some notes: 
1. In the `Demografia` questionnaire, the `[Płec]` here is noted as question instead of the question below it. Because of that, I added a handler in FE where we checked the `header` when it exists. This is how I mapped the first question of Demografia in FE:
```
 {
                                type: "radio",
                                header: "[Płeć / Gender]",
                                text: "Określam swoją płeć jako...",
                                options: [
                                    {
                                        text: "Kobieta",
                                    },
                                    {
                                        text: "Mężczyzna",
                                    },
                                    {
                                        text: "Niebinarny",
                                    },
                                    {
                                        text: "Wolę się samookreślić (odpowiedź otwarta)", 
                                        showTextbox: true,
                                    },
                                    {
                                        text: "Wolę nie mówić",
                                    }
                                ]
                            },
```
You don't have to change it, just for your info XD  Because I also dont know which one is the question lol 

2. I think this one needs confirmation... this question is in the `Emocje (w przerwie)` questionnaire. The BE noted the questionnaire as having 2 questions only, but unless I mapped it wrong, I got 3  :"   The second question got 2 subquestions, and my question is, how will BE handle it? Because now BE stores it as 1 question only
![image](https://user-images.githubusercontent.com/12537724/203556073-4639845c-6721-4ccb-bc75-39ab685487f2.png)

haanh764 commented:
> 2\. estions only, but unless I mapped it wrong, I got 3  :"   T

Hello, thank you for your note. I assume that for (1) it's just for noting the way of mapping between questions in BE - questions in FE, right XD And for (2), because when I implemented it, I didn't know how FE will handle and send the data to BE, then I assumed that it's 1 question and the data send to BE will be some thing like [10 (neg-pos), 40 (ospały-bardzo pobudzony)]. But if FE want to split it, I can change it to 3 questions (?) Wdyt XD

haanh764 commented:
Hello, thank you for your points and notes @teddy-f-47 @artikandri XD I have adjusted the code as below:
1. Adjust the constraints in Answer model. Previously, I thought that (question - answer) per user should be unique, which mean for each question user should have 1 answer only. But then I've changed my mind, because in some case, for example with the Codziennie questionnaire, it's possible for a user to answer the same for the same question everyday. Then I removed the constraint.
2. The user flow to answer/change their answer: so it will query the user's answer by question id and created_date (check if that answer is created today). If for a question, that user have answer it today (there's existed answer in dtb), then he/she can update the answer, otherwise a new answer will be added.

4. I have fixed some issues in the test cases. Now it's oke XD
Please check it again XD

artikandri commented:
> > 2. estions only, but unless I mapped it wrong, I got 3  :"   T
> 
> Hello, thank you for your note. I assume that for (1) it's just for noting the way of mapping between questions in BE - questions in FE, right XD And for (2), because when I implemented it, I didn't know how FE will handle and send the data to BE, then I assumed that it's 1 question and the data send to BE will be some thing like [10 (neg-pos), 40 (ospały-bardzo pobudzony)]. But if FE want to split it, I can change it to 3 questions (?) Wdyt XD

I can try doing the combination, but it will be easier if it's just splitted i think XD   But since the questions are the same, how will BE identify which question is which?

haanh764 commented:
> > > 2. estions only, but unless I mapped it wrong, I got 3  :"   T
> > 
> > 
> > Hello, thank you for your note. I assume that for (1) it's just for noting the way of mapping between questions in BE - questions in FE, right XD And for (2), because when I implemented it, I didn't know how FE will handle and send the data to BE, then I assumed that it's 1 question and the data send to BE will be some thing like [10 (neg-pos), 40 (ospały-bardzo pobudzony)]. But if FE want to split it, I can change it to 3 questions (?) Wdyt XD
> 
> I can try doing the combination, but it will be easier if it's just splitted i think XD But since the questions are the same, how will BE identify which question is which?

I can split it to 2 separated questions - 2 different IDs but in that case I think FE have to make it to be 2 questions somehow different also (I mean 2 headers not 1) :sweat_smile:

artikandri commented:
Hello, I have tried implementing the update api and it returns error like this :")  Did I do something wrong 

![image](https://user-images.githubusercontent.com/12537724/203585549-64e12a63-583c-4530-bce4-7069811df46e.png)
![image](https://user-images.githubusercontent.com/12537724/203585605-6c18d6f2-0f1c-4ad5-af7e-2f1f07812dc4.png)

haanh764 commented:
> Hello, I have tried implementing the update api and it returns error like this :") Did I do something wrong
> 
> ![image](https://user-images.githubusercontent.com/12537724/203585549-64e12a63-583c-4530-bce4-7069811df46e.png) ![image](https://user-images.githubusercontent.com/12537724/203585605-6c18d6f2-0f1c-4ad5-af7e-2f1f07812dc4.png)

hello, I think you have typo in the endpoint url, it is `answers/<int:question_id>`, you missed the `s` :sweat_smile:

artikandri commented:
Hello, I have fixed the typos and now it works well,  thank you for the good work and sorry for the trouble  XD   Approved

teddy-f-47 commented on backend/questionnaires/migrations/0003_initial.py:
```diff
@@ -0,0 +1,257 @@
+        "Jaka liczba tekstów byłaby odpowiednia?",
+        "Co byś zmieniła/zmienił w badaniu?",
+    ]
+    questions = ipip_bfm_questions + demografia_questions +  humor_styles_questions + spane_questions + skala_questions + phq9_questions + pss_questions + zdrowie_questions + ress_ema_questions + paq_questions + zwrotna_questions + sen_questions + stres_rano_questions + stres_wieczor_questions + zdrowie_wieczor_questions + emocje_questions + anotacja_koniec_questions + ankieta_tygodniu_questions
```
Hello! I am sorry I missed this, but I just found out that you missed adding `swlsa_questions` to the questions on line 229 here. So the questions for questionnaire swlsa are not available XD

note: The the `swlsa_questions` should be between `humor_style_questions` and `spane_questions` variables

haanh764 replied:
ah yeah thank you for pointing it out XD I will add it in a second XD

haanh764 replied:
I have updated it XD thank you

teddy-f-47 replied:
thank you! it should be fine now XD
